infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/fsm/sink/RemoteSourceLeadershipManager.java

```diff
@@ -0,0 +1,256 @@
+        try {
+            log.info("adding to the queue {}", event);
+            eventQueue.put(event);
+        } catch (InterruptedException ex) {
+            log.error("Log Replication interrupted Exception: ", ex);
+        }
+    }
+
+    /**
+     * Consumer of the eventQueue.
+     * <p>
+     * This method consumes the log replication events and does the state transition.
+     */
+    private void consume() {
+        try {
+            //  Block until an event shows up in the queue.
+            LogReplicationSinkEvent event = eventQueue.take();
+            processEvent(event);
+
+            communicationFSMConsumer.submit(this::consume);
+
+        } catch (Throwable t) {
+            log.error("Error on event consumer: ", t);
+        }
+    }
+
+    private void processEvent(LogReplicationSinkEvent event) {
+        log.info("processing event {}", event);
+        switch (event.getType()) {
+            case ON_CONNECTION_DOWN:
+                String nodeIdDown = event.getNodeId();
+                log.info("Detected connection down from node={}", nodeIdDown);
+                updateDisconnectedNodes(nodeIdDown);
+                resetRemoteLeader();
+                break;
+            case REMOTE_LEADER_NOT_FOUND:
+                log.info("Remote Leader not found. Retrying...");
+                verifyLeadership();
+                break;
+            case ON_CONNECTION_UP:
+                log.info("Detected connection up from endpoint={}", event.getNodeId());
+                // Add new connected node, for leadership verification
+                updateConnectedNodes(event.getNodeId());
+                verifyLeadership();
+                break;
+            case REMOTE_LEADER_FOUND:
+                log.debug("Remote Leader is found: {}", event.getNodeId());
+                invokeReverseReplication();
+                break;
+            case REMOTE_LEADER_LOSS:
+                log.debug("Remote leader has changed");
+                resetRemoteLeader();
+                verifyLeadership();
+                break;
+            default: {
+                log.warn("Unexpected communication event {}", event.getType());
+            }
+        }
+    }
+
+    private void resetRemoteLeader() {
+        log.debug("Reset remote leader");
+        leaderNodeId = Optional.empty();
+    }
+
+    private void updateConnectedNodes(String nodeId) {
+        this.connectedNodes.add(nodeId);
+    }
+
+    private void updateDisconnectedNodes(String nodeId) {
+        this.connectedNodes.remove(nodeId);
+    }
+
+    public synchronized void setRemoteLeaderNodeId(String leaderId) {
+        log.debug("Set remote leader node id {}", leaderId);
+        leaderNodeId = Optional.ofNullable(leaderId);
+    }
+
+    public synchronized Optional<String> getRemoteLeaderNodeId() {
+        log.trace("Retrieve remote leader node id {}", leaderNodeId);
+        return leaderNodeId;
+    }
+
+    private synchronized void verifyLeadership() {
```

Comment:
this method has the same code as verifyLeadership() in VerifyingRemoteSinkLeaderState.  Can you make both share the code to avoid repetition?  Also, I am thinking if we should consider adding this as a separate state in our existing state machine...

Reply:
I am not very inclined to use the exiting state machine on sink side, as sink really has only 1 state : 
start(query Leadership) and end (reverseReplicate).

Reply:
Will try to avoid the repetition and have the code in a common place